cider/tests/functionality/CiderSemiJoinTest.cpp

```diff
@@ -0,0 +1,140 @@
+        "BOOLEAN, r_f INTEGER);";
+    resetHashTable();
+  }
+
+  void resetHashTable() override {
+    build_table_.reset(new CiderBatch(QueryDataGenerator::generateBatchByTypes(
+        100,
+        {"r_a", "r_b", "r_c", "r_d", "r_e", "r_f"},
+        {CREATE_SUBSTRAIT_TYPE(I64),
+         CREATE_SUBSTRAIT_TYPE(I32),
+         CREATE_SUBSTRAIT_TYPE(Fp64),
+         CREATE_SUBSTRAIT_TYPE(Fp32),
+         CREATE_SUBSTRAIT_TYPE(Bool),
+         CREATE_SUBSTRAIT_TYPE(I32)})));
+    duckDbQueryRunner_.createTableAndInsertData(
+        build_table_name_, build_table_ddl_, {build_table_});
+  }
+};
+
+class CiderSemiJoinNullableTest : public CiderJoinTestBase {
+ public:
+  CiderSemiJoinNullableTest() {
+    table_name_ = "table_probe";
+    create_ddl_ =
+        "CREATE TABLE table_probe(l_a BIGINT, l_b INTEGER, l_c DOUBLE, l_d FLOAT, l_e "
+        "BOOLEAN);";
+    input_ = {std::make_shared<CiderBatch>(
+        QueryDataGenerator::generateBatchByTypes(10,
+                                                 {"l_a", "l_b", "l_c", "l_d", "l_e"},
+                                                 {CREATE_SUBSTRAIT_TYPE(I64),
+                                                  CREATE_SUBSTRAIT_TYPE(I32),
+                                                  CREATE_SUBSTRAIT_TYPE(Fp64),
+                                                  CREATE_SUBSTRAIT_TYPE(Fp32),
+                                                  CREATE_SUBSTRAIT_TYPE(Bool)}))};
+
+    build_table_name_ = "table_hash";
+    build_table_ddl_ =
+        "CREATE TABLE table_hash(r_a BIGINT, r_b INTEGER, r_c DOUBLE, r_d FLOAT, r_e "
+        "BOOLEAN, r_f INTEGER);";
+    resetHashTable();
+  }
+
+  void resetHashTable() override {
+    build_table_.reset(new CiderBatch(QueryDataGenerator::generateBatchByTypes(
+        3,
+        {"r_a", "r_b", "r_c", "r_d", "r_e", "r_f"},
+        {CREATE_SUBSTRAIT_TYPE(I64),
+         CREATE_SUBSTRAIT_TYPE(I32),
+         CREATE_SUBSTRAIT_TYPE(Fp64),
+         CREATE_SUBSTRAIT_TYPE(Fp32),
+         CREATE_SUBSTRAIT_TYPE(Bool),
+         CREATE_SUBSTRAIT_TYPE(I32)},
+        {2, 2, 2, 2, 2, 2})));
+    duckDbQueryRunner_.createTableAndInsertData(
+        build_table_name_, build_table_ddl_, {build_table_});
+  }
+};
+
+TEST_F(CiderSemiJoinTest, semi) {
+  std::string sql_semi = "SELECT l_a FROM table_probe SEMI JOIN table_hash ON l_a = r_a";
+  std::string sql_semi_in =
+      "SELECT l_a FROM table_probe WHERE l_a IN (SELECT r_a FROM table_hash)";
+  assertJoinQueryAndReset(sql_semi_in, "semi.json");
+}
+
+TEST_F(CiderSemiJoinTest, anti) {
+  std::string sql_anti = "SELECT l_a FROM table_probe ANTI JOIN table_hash ON l_a = r_a";
+  std::string sql_anti_in =
+      "SELECT l_a FROM table_probe WHERE l_a NOT IN (SELECT r_a FROM table_hash)";
+  assertJoinQueryAndReset(sql_anti_in, "anti.json");
+}
+
+TEST_F(CiderSemiJoinNullableTest, semi) {
+  std::string sql_semi = "SELECT l_a FROM table_probe SEMI JOIN table_hash ON l_a = r_a";
```

Comment:
`sql_semi` not used?

Reply:
yes, just want to show they are same semantic.

Reply:
Maybe add comments at file head to address the same semantics instead of using these lines in several cases?

Reply:
Does duckdb support ```sql_semi``` and ```sql_anti``` ?

Reply:
duckdb does not support key word `semi` or `anti`, but its parser/planner will recognize IN clause as semi/anti join pattern.

Reply:
> Maybe add comments at file head to address the same semantics instead of using these lines in several cases?

updated.